refactor(http-client-csharp): dump deprecated `description` properties

archerzz opened the pull request:
- emitter: replace all `description` properties of input types with `summary` and `doc`
- generator:
  - update json converters for the emitter schema change
  - only expose `Summary` and `Doc` properties if they are used by other classes
- update test cases accordingly

part of #4771

azure-sdk commented:
**API change check**

API changes are not detected in this pull request.

ArcturusZhang commented on packages/http-client-csharp/emitter/src/lib/typespec-server.ts:
```diff
@@ -73,7 +74,7 @@ export function resolveServers(
         Name: "host",
         NameInRequest: "host",
-        Description: server.description,
+        Doc: server.description,
```
why the server only has description? should this also be changed to summary and doc?

archerzz replied:
this is from typespec library: https://github.com/microsoft/typespec/blob/1cf8601d0f65f707926d58d56566fb0cb4d4f4ff/packages/http/src/decorators.ts#L464-L468

there is only `description`

ArcturusZhang commented on packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/test/Providers/ClientProviders/ClientProviderSubClientTests.cs:
```diff
@@ -14,11 +14,11 @@ namespace Microsoft.Generator.CSharp.ClientModel.Tests.Providers.ClientProviders
-        private static readonly InputClient _catClient = new("cat", "CatClient description", [], [], _animalClient.Name);
-        private static readonly InputClient _hawkClient = new("hawkClient", "HawkClient description", [], [], _animalClient.Name);
-        private static readonly InputClient _huskyClient = new("husky", "HuskyClient description", [], [], _dogClient.Name);
+        private static readonly InputClient _animalClient = new("animal", "", "AnimalClient description", [], [], TestClientName);
+        private static readonly InputClient _dogClient = new("dog", "", "DogClient description", [], [], _animalClient.Name);
+        private static readonly InputClient _catClient = new("cat", "", "CatClient description", [], [], _animalClient.Name);
+        private static readonly InputClient _hawkClient = new("hawkClient", "", "HawkClient description", [], [], _animalClient.Name);
+        private static readonly InputClient _huskyClient = new("husky", "", "HuskyClient description", [], [], _dogClient.Name);
```
i think we should do `string.Empty` instead of `""`.
Also should this be `null`?

ArcturusZhang replied:
I think it is never required to have summary or doc there these two properties should be nullable.

archerzz replied:
> I think it is never required to have summary or doc there these two properties should be nullable. And there should be difference between when we have:
> 
> ```
> model Foo {}
> ```
> 
> and
> 
> ```
> @doc("")
> model Foo {}
> ```

Zero values are the same here. You don't want to generate an empty doc anyway. Also, we're not consistent in our code base. Some defines `string?` and others define `string`. I just follow what the current codes were written.

ArcturusZhang replied:
This comment follows this one: https://github.com/microsoft/typespec/pull/5154#discussion_r1851235836
For example an input enum type is not required to have a doc or summary or description, therefore our definition here is wrong in the first place.
We define this in this way, because previously these are defined for azure purpose, and for azure libraries, types, properties are not allowed without documentation. But in unbranded libraries, there is no such rules.

Azure part should change accordingly if we change it here - assign default value when an output type is contructed.

archerzz replied:
changed all `Summary` and `Doc` as `string?`

ArcturusZhang commented on packages/http-client-csharp/generator/Microsoft.Generator.CSharp.Input/src/InputTypes/InputEnumType.cs:
```diff
@@ -7,13 +7,13 @@ namespace Microsoft.Generator.CSharp.Input
             Accessibility = accessibility;
             Deprecated = deprecated;
-            Description = description;
+            Description = string.IsNullOrEmpty(summary) ? (doc ?? string.Empty) : summary;
```
I think we should change this property to two new properties `Summary` and `Doc`
The classes in input types should just be the C# version of those typescript interfaces.

archerzz replied:
I see two different strategies regarding the input types between `http-client-csharp` and `autorest.csharp`:
1. In `http-client-csharp`, we declare them as `class`. So, I keep the minimal interface here.
2. In `autorest.csharp`, we declare them as `record`. So, I keep the interface like the original data schema as much as possible.

ArcturusZhang replied:
record is a class, there is no differences.
Making this change might take quite a few efforts, this I admit.
But we should not be avoiding them - I have done multiple of those when consolidating the models part.

archerzz commented:
Changes since last review:
- Change all input types to make `Summary` and `Doc` properties as `string?`
- Update type converters to **not** set default value during de-serialization. Set them when generating output models.
- Some minor fixes and refactoring

chunyu3 commented on packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/ScmMethodProviderCollection.cs:
```diff
@@ -424,7 +425,7 @@ private ScmMethodProvider BuildProtocolMethod(MethodProvider createRequestMethod
             var methodSignature = new MethodSignature(
                 isAsync ? _cleanOperationName + "Async" : _cleanOperationName,
-                FormattableStringHelpers.FromString(Operation.Description),
+                DocHelpers.GetFormattableDescription(Operation.Summary, Operation.Doc) ?? FormattableStringHelpers.FromString(Operation.Name),
```
Description of MethodSignature is nullable. Why do we mock the operation.Name as its description here?

archerzz replied:
> Description of MethodSignature is nullable. Why do we mock the operation.Name as its description here?

It's because previously we use `Operation.Name` if `Description` is not available: https://github.com/microsoft/typespec/blob/b6dcb0482055a182b2545803b4c8d618fca6cfe8/packages/http-client-csharp/generator/Microsoft.Generator.CSharp.Input/src/InputTypes/Serialization/TypeSpecInputOperationConverter.cs#L91

After discussion with Dapeng, we thought we should do less in the input type converters. So, I moved the logic in `TypeSepcInputOperationConverter` to the output model here.

We have similar changes elsewhere in this PR as well.

azure-sdk commented:
<!--chronus-github-change-commenter-->
:x: There is undocummented changes. Run `chronus add` to add a changeset or [click here](https://github.com/archerzz/typespec/new/http-client-csharp/adopt-tcgc-doc-properties?filename=.chronus/changes/http-client-csharp-adopt-tcgc-doc-properties-2024-11-17-7-36-3.md&value=---%0A%23%20Change%20versionKind%20to%20one%20of%3A%20internal%2C%20fix%2C%20dependencies%2C%20feature%2C%20deprecation%2C%20breaking%0AchangeKind%3A%20fix%0Apackages%3A%0A%20%20-%20typespec-vscode%0A---%0A%0Arefactor(http-client-csharp)%3A%20dump%20deprecated%20%60description%60%20properties).

**The following packages have changes but are not documented.**
 - :x:`typespec-vscode`

<details>
<summary>Show changes</summary>


</details>

azure-sdk commented:
<!-- _TSP_TRYIT_COMMENT_ -->
You can try these changes here

| [🛝 Playground]( https://cadlplayground.z22.web.core.windows.net/prs/5154/)|[🌐 Website](https://tspwebsitepr.z22.web.core.windows.net/prs/5154/)|[📚 Next docs](https://tspwebsitepr.z22.web.core.windows.net/prs/5154/docs/next.html) |
| ---|---|--- |
